### PR TITLE
Untangling list provider from logic and fixing big endian safety bug

### PR DESCRIPTION
**Pull request by robertbastian**
The bug was that databake baked little-endian bytes. It now bakes both with a `cfg` to choose the correct one.

This also fixes a second bug where on BE systems we wouldn't read (and discard) the DFA bytes, leading to a parse error.

### CODE REVIEW

File: components/list/src/patterns.rs [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+use regex_automata::dfa::Automaton;
+use writeable::{LengthHint, Writeable};
+
+impl<'data> ListFormatterPatternsV1<'data> {
+    /// Creates a new [`ListFormatterPatternsV1`] from the given patterns. Fails if any pattern is invalid.
+    ///
+    /// See [`ListJoinerPattern::from_str`]. `allow_prefix` will be true for `pair` and `end` patterns,
+    /// `allow_suffix` for `start` and `pair` patterns.
+    #[cfg(feature = "datagen")]
+    pub fn try_new(
```
Comment (Manishearth) [inline components/list/src/patterns.rs:20]: suggestion: should these just be twelve parameters?

Comment (robertbastian) [reply components/list/src/patterns.rs:20]: Yes, but that'd be a breaking change now.

Comment (Manishearth) [reply components/list/src/patterns.rs:20]: ah

**Review by sffc — COMMENTED** [review]
LGTM; thanks for finding and fixing this!

File: components/list/src/provider/serde_dfa.rs [hunk 2]
```diff
@@ -52,58 +63,57 @@ impl serde::Serialize for StringMatcher<'_> {
...
-                dfa_bytes: Cow::Borrowed(&[]),
-                pattern: None,
-            });
+            return Ok(None);
         }
 
-        let dfa_bytes = <CowBytesWrap<'de>>::deserialize(deserializer)?.0;
-
         // Verify safety invariant
         DFA::from_bytes(&dfa_bytes).map_err(|e| {
```
Comment (sffc) [inline components/list/src/provider/serde_dfa.rs:103]: Observation: You could envision a BE system with baked data providing a data service to LE clients. In this case, the DFA validator will run with improper data. I assume we will get the expected behavior of these data structs simply failing to deserialize.

Comment (robertbastian) [reply components/list/src/provider/serde_dfa.rs:103]: > BE system with baked data providing a data service to LE clients

How? BE and LE systems cannot share memory. They will have to go through serde.

Comment (robertbastian) [reply components/list/src/provider/serde_dfa.rs:103]: (and going through serde serializes as LE, not native, so that works from BE systems)

**Pull request merged**